docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,210 @@
+
+#### typelib
+All attributes of newly created typelib are initialized with empty tables / arrays and `nil`. For instance, `typelib.newtable()` initializes its properties with an empty table and index / index result type as `nil`. Additionally, all arguments are passed by references.
+
+| Instance Attributes | Type | Description |
+| ------------- | ------------- | ------------- |
+| `niltype` | `typelib` | an immutable runtime representation of the built-in type `nil` |
+| `unknown` | `typelib` | an immutable runtime representation of the built-in type `unknown` |
+| `never` | `typelib` | an immutable runtime representation of the built-in type `never` |
+| `any` | `typelib` | an immutable runtime representation of the built-in type `any` |
+| `boolean` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
+| `number` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
+| `string` | `typelib` | returns an immutable runtime representation of the built-in type `string` |
+
+| Instance Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is syntactically a subtype or equal to arg in the type hierarchy |
+| `__eq(arg: typelib)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg in the type hierarchy |
+
+| Static Methods | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
+| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
+| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
+| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
+| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
+| `newtable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?)` | `typelib` | returns a mutable runtime representation of a `table` type |
+| `newmetatable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?, metatable: typelib?)` | `typelib` | returns a mutable runtime representation of a metatable. `metatable` argument needs to be the same type as `typelib.newtable()` |
+| `newfunction(parameters: {typelib} \| typelib?, returns: {typelib} \| typelib?)` | `typelib` | returns a mutable runtime representation of a `function` type. Calling `newfunction(X)` will by default set `parameters` to `X` |
```

Comment:
Alternative design space is

```luau
typelib.create("literal", "my string")
typelib.create("literal", true)
typelib.create("union", {my types})
typelib.create("table", props, indexer)
typelib.create("function", params, rets)
```

Reply:
yeah, separate constructors vs one overloaded constructor. I personally prefer separate constructors.

Reply:
i would prefer separate constructors as well. for my project, i would like to keep it that way 🙂